### PR TITLE
HDDS-6528. Adding Unit-Test cases for S3-Gateway Bucket-Endpoint Metrics.

### PR DESCRIPTION
**Pull request by ArafatKhan2198**
## What changes were proposed in this pull request?

Added Unit Tests for Bucket Endpoint Metrics for S3-Gateway 

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-6528

## How was this patch tested?

Wrote Unit tests for existing metrics 

@jojochuang @aryangupta1998 @siddhantsangwan @adoroszlai

### CODE REVIEW

File: hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/metrics/TestS3GatewayMetrics.java [hunk 1]
```diff
@@ -110,4 +123,146 @@ public void testHeadObject() throws Exception {
...
+          null, null, 1000, null,
+          null, null, "random", null,
+          null, null);
+      fail();
+    } catch (OS3Exception e) {
+    }
+
+    curMetric = metrics.getGetBucketFailure();
+    assertEquals(1L, curMetric - oriMetric);
+  }
```
Comment (adoroszlai) [inline hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/metrics/TestS3GatewayMetrics.java:156]: Can you please add separate test cases (methods) for success/failure scenarios instead of bundling them in the same?

**Review by adoroszlai — CHANGES_REQUESTED** [review]
Thanks @ArafatKhan2198 for working on this.  There is compile error due to recent change on `master` (HDDS-6466).  Can you please merge `origin/master` into your branch?

File: hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/metrics/TestS3GatewayMetrics.java [hunk 2]
```diff
@@ -110,4 +123,146 @@ public void testHeadObject() throws Exception {
     long curMetric = metrics.getHeadKeySuccess();
     assertEquals(1L, curMetric - oriMetric);
   }
-}
+
+  @Test
+  public void testGetBucketEndpointMetric() throws Exception {
```
[the remark this answers is not in the document]
Comment (ArafatKhan2198) [reply hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/metrics/TestS3GatewayMetrics.java:128]: The reason for that is because these Unit tests are written for all the metrics associated with the Endpoints exposed by the S3 API! would you want me to remove it?

[the remark this answers is not in the document]
Comment (adoroszlai) [reply hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/metrics/TestS3GatewayMetrics.java:128]: Yes, please follow naming of existing test methods like `testHeadBucket`, `testListBucket`, `testHeadObject`.

[the remark this answers is not in the document]
Comment (ArafatKhan2198) [reply hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/metrics/TestS3GatewayMetrics.java:128]: Sure will do!!

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java [hunk 3]
```diff
@@ -155,6 +155,10 @@ public Response get(
       } else {
         throw ex;
       }
+    } catch (OS3Exception ex) {
+      getMetrics().incGetBucketFailure();
+      LOG.error("Bucket Does not Exist " + bucketName, ex);
```
[the remark this answers is not in the document]
Comment (ArafatKhan2198) [reply hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java:160]: You are absolutely right I will replace Log.error() with Log.debug() !!

File: hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/metrics/TestS3GatewayMetrics.java (hunk 1 above)
Comment (adoroszlai) [reply hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/metrics/TestS3GatewayMetrics.java]: > Each method is associated with a single S3 Endpoint and there are 2 metrics associated with each endpoint, hence we are measuring the number of times a method succeeds and fails, and once an object is initialised we can reuse it for both the metrics

Currently each test method is associated with a method on one of the endpoints.  E.g. `testCreateBucket` tests `bucketEndpoint.put()`, while `testDeleteBucket` tests `bucketEndpoint.delete()`.  If object reuse is so important, why not merge all these together in a huge `testBucketEndpoint` method?

The reasons for separating these are at least:

 1. Small, self-contained test methods are easier to understand.
 2. Independence: test cases may pass or fail regardless of the result of other test cases.  If multiple cases are bundled in the same method, failure of one case may hide the result of the subsequent case.

We can apply the same reasons to testing different metrics on the same endpoint and method pairs.

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java [hunk 4]
```diff
@@ -145,6 +145,10 @@ public Response get(
       } else {
         throw ex;
       }
+    } catch (OS3Exception ex) {
+      getMetrics().incGetBucketFailure();
+      LOG.debug("Bucket Does not Exist " + bucketName, ex);
+      throw newError(S3ErrorTable.NO_SUCH_BUCKET, bucketName, ex);
```
Comment (adoroszlai) [inline hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java:151]: We should just rethrow `ex`, because:

 * `OS3Exception` already has an error code and proper message.  No need to wrap in another exception.
 * Cause may be different, not necessarily `NO_SUCH_BUCKET`.

On second thought we also don't need the debug message, because the exception is logged at debug level in `newError`.

https://github.com/apache/ozone/blob/e81ab0a3f76b09072c23e0151c5e28673623ed6c/hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/exception/S3ErrorTable.java#L141-L145

```suggestion
      throw ex;
```

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java [hunk 5]
```diff
@@ -407,6 +411,10 @@ public S3BucketAcl getAcl(String bucketName)
         LOG.error("Failed to get acl of Bucket " + bucketName, ex);
         throw newError(S3ErrorTable.INTERNAL_ERROR, bucketName, ex);
       }
+    } catch (OS3Exception ex) {
+      getMetrics().incGetAclFailure();
+      LOG.debug("Failed to get acl of Bucket " + bucketName, ex);
+      throw newError(S3ErrorTable.NO_SUCH_BUCKET, bucketName, ex);
```
Comment (adoroszlai) [inline hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java:417]: Same here:

```suggestion
      throw ex;
```

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java [hunk 6]
```diff
@@ -507,6 +514,10 @@ public Response putAcl(String bucketName, HttpHeaders httpHeaders,
       LOG.error("Error in set ACL Request for bucket: {}", bucketName,
           exception);
       throw exception;
+    } catch (OS3Exception ex) {
+      getMetrics().incPutAclFailure();
+      LOG.debug("Failed to put acl of Bucket " + bucketName, ex);
```
Comment (adoroszlai) [inline hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java:519]: We don't need the debug message, because the exception is logged at debug level in `newError`.

https://github.com/apache/ozone/blob/e81ab0a3f76b09072c23e0151c5e28673623ed6c/hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/exception/S3ErrorTable.java#L141-L145

```suggestion
```

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/metrics/S3GatewayMetrics.java [hunk 7]
```diff
@@ -44,7 +44,9 @@
 
   // BucketEndpoint
   private @Metric MutableCounterLong getBucketSuccess;
+
```
Comment (adoroszlai) [inline hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/metrics/S3GatewayMetrics.java:47]:
```suggestion
```

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/metrics/S3GatewayMetrics.java [hunk 8]
```diff
@@ -44,7 +44,9 @@
...
   private @Metric MutableCounterLong getBucketFailure;
+
```
Comment (adoroszlai) [inline hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/metrics/S3GatewayMetrics.java:49]:
```suggestion
```

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java [hunk 9]
```diff
@@ -240,7 +244,7 @@ public Response put(@PathParam("bucket") String bucketName,
       if (exception.getResult() == ResultCodes.INVALID_BUCKET_NAME) {
         throw newError(S3ErrorTable.INVALID_BUCKET_NAME, bucketName, exception);
       }
-      LOG.error("Error in Create Bucket Request for bucket: {}", bucketName,
+      LOG.debug("Error in Create Bucket Request for bucket: {}", bucketName,
```
Comment (adoroszlai) [inline hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java:247]: Nit: unrelated change.

```suggestion
      LOG.error("Error in Create Bucket Request for bucket: {}", bucketName,
```

**Review by adoroszlai — COMMENTED** [review]
Thanks @ArafatKhan2198 for updating the patch.  On second look I've found a few more issues.

File: hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/metrics/TestS3GatewayMetrics.java (hunk 1 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/metrics/TestS3GatewayMetrics.java]: Yes you are absolutely right on that part, I'll make the necessary changes
Thanks for the help !!

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java (hunk 4 above)
Comment (jojochuang) [reply hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java:151]: +1 we should just increment the counter and retrhwo

File: hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/metrics/TestS3GatewayMetrics.java [hunk 10]
```diff
@@ -110,4 +123,146 @@ public void testHeadObject() throws Exception {
...
+    curMetric = metrics.getGetAclFailure();
+    assertEquals(1L, curMetric - oriMetric);
+  }
+
+  @Test
+  public void testPutAcl() throws Exception {
+    long oriMetric = metrics.getPutAclSuccess();
+
+    clientStub.getObjectStore().createS3Bucket("b1");
+    InputStream inputBody = TestBucketAcl.class.getClassLoader()
```
Comment (jojochuang) [inline hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/metrics/TestS3GatewayMetrics.java:238]: close this input stream at the end of the test.

File: hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/metrics/TestS3GatewayMetrics.java [hunk 11]
```diff
@@ -110,4 +123,146 @@ public void testHeadObject() throws Exception {
...
+      bucketEndpoint.put("unknown_bucket", ACL_MARKER, headers, inputBody);
+      fail();
+    } catch (OS3Exception ex) {
+    }
+    curMetric = metrics.getPutAclFailure();
+    assertEquals(1L, curMetric - oriMetric);
+  }
+
+  private OzoneClient createClientWithKeys(String... keys) throws IOException {
+    OzoneClient client = new OzoneClientStub();
```
[the remark this answers is not in the document]
Comment (ArafatKhan2198) [reply hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/metrics/TestS3GatewayMetrics.java:259]: I am sorry I forgot to remove it !! 
will change it, thanks !!

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java (hunk 4 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java:151]: Sure will do !!

File: hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/metrics/TestS3GatewayMetrics.java (hunk 10 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/metrics/TestS3GatewayMetrics.java:238]: Thanks will do !!

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/metrics/S3GatewayMetrics.java (hunk 8 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/metrics/S3GatewayMetrics.java:49]: Will make the change, thanks !!

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java (hunk 6 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java:519]: Will make the change, thanks !!

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java (hunk 5 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java:417]: Will make the change, thanks !!

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/metrics/S3GatewayMetrics.java [hunk 12]
```diff
@@ -317,4 +317,52 @@ public long getHeadBucketSuccess() {
...
+  public long getPutAclFailure() {
+    return putAclFailure.value();
+  }
+
+  public long getListMultipartUploadsSuccess() {
+    return listMultipartUploadsSuccess.value();
+  }
+
+  public long getListMultipartUploadsFailure() {
+    return listMultipartUploadsSuccess.value();
```
Comment (adoroszlai) [inline hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/metrics/S3GatewayMetrics.java:366]: Sorry, I just noticed this typo:

```suggestion
  public long getListMultipartUploadsFailure() {
    return listMultipartUploadsFailure.value();
```

File: hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/metrics/TestS3GatewayMetrics.java [hunk 13]
```diff
@@ -110,4 +123,160 @@ public void testHeadObject() throws Exception {
...
+    clientStub.getObjectStore().createS3Bucket("b1");
+    InputStream inputBody = TestBucketAcl.class.getClassLoader()
+        .getResourceAsStream("userAccessControlList.xml");
+
+    try {
+      bucketEndpoint.put("unknown_bucket", ACL_MARKER, headers, inputBody);
+      fail();
+    } catch (OS3Exception ex) {
+    }
+    inputBody.close();
```
Comment (adoroszlai) [inline hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/metrics/TestS3GatewayMetrics.java:270]: Nit: using try-with-resources we can ensure that the stream is closed even in case of failure.

```suggestion
    try (InputStream inputBody = TestBucketAcl.class.getClassLoader()
          .getResourceAsStream("userAccessControlList.xml")) {
      bucketEndpoint.put("unknown_bucket", ACL_MARKER, headers, inputBody);
      fail();
    } catch (OS3Exception ex) {
    }
```

File: hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/metrics/TestS3GatewayMetrics.java [hunk 14]
```diff
@@ -110,4 +123,160 @@ public void testHeadObject() throws Exception {
...
+    long curMetric = metrics.getPutAclSuccess();
+    assertEquals(1L, curMetric - oriMetric);
+  }
+
+  @Test
+  public void testPutAclFailure() throws Exception {
+    // Failing the putACL endpoint by applying ACL on a non-Existent Bucket
+    long oriMetric = metrics.getPutAclFailure();
+
+    clientStub.getObjectStore().createS3Bucket("b1");
```
Comment (adoroszlai) [inline hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/metrics/TestS3GatewayMetrics.java:261]: Nit: `b1` bucket is not needed for this test case.

File: hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/metrics/TestS3GatewayMetrics.java [hunk 15]
```diff
@@ -110,4 +123,160 @@ public void testHeadObject() throws Exception {
...
+  @Test
+  public void testPutAclSuccess() throws Exception {
+    long oriMetric = metrics.getPutAclSuccess();
+
+    clientStub.getObjectStore().createS3Bucket("b1");
+    InputStream inputBody = TestBucketAcl.class.getClassLoader()
+        .getResourceAsStream("userAccessControlList.xml");
+
+    bucketEndpoint.put("b1", ACL_MARKER, headers, inputBody);
+    inputBody.close();
```
Comment (adoroszlai) [inline hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/metrics/TestS3GatewayMetrics.java:251]: Nit: using try-with-resources we can ensure that the stream is closed even in case of failure.

```suggestion
    try (InputStream inputBody = TestBucketAcl.class.getClassLoader()
          .getResourceAsStream("userAccessControlList.xml")) {

      bucketEndpoint.put("b1", ACL_MARKER, headers, inputBody);
    }
```

**Review by adoroszlai — COMMENTED** [review]
Thanks @ArafatKhan2198 for updating the patch, especially for separating the test cases.  I noticed one typo where the wrong metric is returned, otherwise it's looks good to me.

**Review by adoroszlai — APPROVED** [review]

**Review by jojochuang — APPROVED** [review]
